contracts/core/RebalancingSet.sol

```diff
@@ -0,0 +1,389 @@
+        address _initialSet,
+        uint256 _initialShareRatio,
+        uint256 _proposalPeriod,
+        uint256 _rebalanceInterval,
+        string _name,
+        string _symbol
+    )
+        public
+        DetailedERC20(_name, _symbol, 18)
+    {
+        // Require day long proposal period
+        require(_proposalPeriod > 86400);
+
+        // Require one day between end of rebalance and proposing another rebalance
+        require(_rebalanceInterval > 86400);
+
+        factory = _factory;
+        manager = _manager;
+        currentSet = _initialSet;
+        shareRatio = _initialShareRatio;
+
+        proposalPeriod = _proposalPeriod;
+        rebalanceInterval = _rebalanceInterval;
+        lastRebalanceTimestamp = block.timestamp;
+        rebalanceState = State.Default;
+    }
+
+    /* ============ Public Functions ============ */
+
+    /**
+     * Function used to set the terms of the next rebalance and start the proposal period
+     *
+     *
+     * @param _rebalancingSet               The Set to rebalance into
+     * @param _auctionLibrary               The library used to calculate the Dutch Auction price
+     * @param _curveCoefficient             The slope (or convexity) of the price curve
+     * @param _auctionPriceDivisor          The granularity with which the prices change
+     * @param _auctionStartPrice            The price to start the auction at
+     */
+    function propose(
+        address _rebalancingSet,
+        address _auctionLibrary,
+        uint256 _curveCoefficient,
+        uint256 _auctionStartPrice,
+        uint256 _auctionPriceDivisor
+    )
+        external
+    {
+        // Make sure it is manager that is proposing the rebalance
+        require(msg.sender == manager);
+
+        // New proposal cannot be made during a rebalance period
+        require(rebalanceState == State.Default);
+
+        // Make sure enough time has passed from last rebalance to start a new proposal
+        require(block.timestamp >= lastRebalanceTimestamp.add(rebalanceInterval));
+
+        // Set auction parameters
+        rebalancingSet = _rebalancingSet;
+        auctionLibrary = _auctionLibrary;
+        curveCoefficient = _curveCoefficient;
+        auctionStartPrice = _auctionStartPrice;
+        auctionPriceDivisor = _auctionPriceDivisor;
+        remainingCurrentSets = shareRatio.mul(totalSupply_).div(10**18);
+
+        // Update state parameters
+        proposalStartTime = block.timestamp;
+        rebalanceState = State.Proposal;
+
+        emit RebalanceProposed(
+            _rebalancingSet,
+            _auctionLibrary,
+            proposalStartTime.add(proposalPeriod)
+        );
+    }
+
+    /*
+     * Initiate rebalance for the rebalancing set. Users can now submit bids.
+     *
+     */
+    function rebalance()
+        external
+    {
+        // Must be in "Proposal" state before going into "Rebalance" state
+        require(rebalanceState == State.Proposal);
```

Comment:
We could probably abstract the state transition and timing checks into a private function

Reply:
Can you explain what you mean? If it's what I think you mean I suppose that would look cleaner but not sure it'd be more efficient.